File: csharp/CodeGeneration/Services/Github/Github.Service.cs
/*
  * 
 * The MIT License (MIT)
 * 
 * Copyright (c) [2023] [oscar garcia]
 * 
 * Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated documentation files (the "Software"),
 * to deal in the Software without restriction, including without limitation the rights to use, copy, modify, merge, publish, distribute, sublicense,
 * and/or sell copies of the Software, and to permit persons to whom the Software is furnished to do so, subject to the following conditions:
 * 
 * The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software.
 * 
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 *
 * Author: oscar garcia - ozkary
 */
namespace CodeGeneration.Services.GitHub;

using System;
using System.Collections.Generic;
using System.Net.Http;
using System.Text.Json;
using System.Net.Http.Headers;
using System.Threading.Tasks;
using System.Text;

/// <summary>
/// GitHub API service
/// </summary>
internal class GitHubService
{
    private const string RepoUrl = "https://api.github.com/repos/";
    private const string IssuesRoute = "issues";
    private const string CommentsRoute = "comments";

    const string UserAgent = "Ozkary OpenAI V1.0.0";

    private static string BuildQueryString(Dictionary<string, string> parameters)
    {
        var queryString = new List<string>();
        foreach (var parameter in parameters)
        {
            queryString.Add($"{Uri.EscapeDataString(parameter.Key)}={Uri.EscapeDataString(parameter.Value)}");
        }
        return string.Join("&", queryString);
    }

    /// <summary>
    /// Get issues from a GitHub repository
    /// </summary>
    /// <param name="route">The GitHub repository route</param>
    /// <param name="params">The GitHub API parameters</param>
    /// <param name="accessToken">The GitHub access token</param>
    /// <returns>A list of issues</returns>
    /// <exception cref="ArgumentNullException">Thrown when the access token is null or empty</exception>
    /// <exception cref="HttpRequestException">Thrown when the request fails</exception>
    /// <exception cref="JsonException">Thrown when the response body can't be parsed</exception>
    /// <exception cref="InvalidOperationException">Thrown when the response is not successful</exception>
    /// <exception cref="Exception">Thrown when an unknown error occurs</exception>
    /// <remarks>
    /// See the GitHub API documentation for more information:
    /// https://docs.github.com/en/rest/reference/issues#list-repository-issues
    /// </remarks>
   public static async Task<List<Issue>> GetIssues(string route, Parameter @params, string? accessToken)
    {
        if (String.IsNullOrEmpty(accessToken)){
            throw new ArgumentNullException(nameof(accessToken));
        }

        // Set up HttpClient and authentication if needed
        using (HttpClient httpClient = new HttpClient())
        {
            if (!string.IsNullOrEmpty(accessToken))
            {
                Console.WriteLine("Setting up authentication for GitHub API request");
                // httpClient.DefaultRequestHeaders.Authorization = new AuthenticationHeaderValue("Bearer", accessToken);
                httpClient.DefaultRequestHeaders.Add("Authorization", $"Bearer {accessToken}");
                httpClient.DefaultRequestHeaders.Add("User-Agent",UserAgent);

            }

            // Send the API request
            string apiUrl = $"{RepoUrl}{route}/{IssuesRoute}";

            Console.WriteLine($"Sending request to {apiUrl}");

            var apiParams = new Dictionary<string, string>
            {
                { "labels", @params.Label },
                { "state", @params.State }
            };

            var queryString = BuildQueryString(apiParams);
            HttpResponseMessage response = await httpClient.GetAsync(apiUrl + "?" + queryString);

            List<Issue> issuesList = new List<Issue>();             

            if (response.IsSuccessStatusCode)
            {
                // Parse the response body                                  
                string responseBody = await response.Content.ReadAsStringAsync();
                if (!String.IsNullOrEmpty(responseBody)){
                    List<Issue>? issues = JsonSerializer.Deserialize<List<Issue>>(responseBody);
                    if (issues != null){
                        issuesList.AddRange(issues);
                    }         
                    Console.WriteLine($"Issues found: {issuesList.Count}");
                }                                              
            }
            else
            {
                Console.WriteLine($"Error getting the issues: {response.StatusCode} - {response.ReasonPhrase}");
            }

            return issuesList;
        }
    }        

    public static async Task<bool> PostIssueComment(string repo, int issueId, string comment, string accessToken) {
        // Set up HttpClient and authentication if needed
        using (HttpClient httpClient = new HttpClient())
        {
            if (!string.IsNullOrEmpty(accessToken))
            {
                httpClient.DefaultRequestHeaders.Add("Authorization", $"Bearer {accessToken}");
                httpClient.DefaultRequestHeaders.Add("User-Agent",UserAgent);
            }

            // Send the API request
            string apiUrl = $"{RepoUrl}{repo}/{IssuesRoute}/{issueId}/{CommentsRoute}";

            Console.WriteLine($"Sending request to {apiUrl}");

            // var body = new Dictionary<string, string>
            // {
            //     { "body", comment }
            // };

            // Create the json comment payload
            var payload = new
            {
                body = comment
            };

            string jsonPayload = JsonSerializer.Serialize(payload);

            // Set the Content-Type header and create the StringContent object
            StringContent stringContent = new StringContent(jsonPayload, Encoding.UTF8, "application/json");                        
            HttpResponseMessage response = await httpClient.PostAsync(apiUrl, stringContent);

            if (response.IsSuccessStatusCode)
            {
                return true;
            }
            else
            {
                Console.WriteLine($"Error: {response.StatusCode} - {response.ReasonPhrase}");
                return false;
            }
        }

    }
}

File: python/services/github/github_service.py
"""GitHub API service."""
# -*- coding: utf-8 -*-
#
#  2023 ozkary.com.
#
#  Github service
#

from requests import get, post
from typing import List
from services.github.github_types import Issue, Parameter

class GitHubService:
    
    repo_url = "https://api.github.com/repos/"
    issues_route = "issues"
    comments_route = "comments"

    # Send a request to the GitHub API to retrieve the issues from a repo
    @staticmethod
    def get_github_issues(route: str, params: Parameter = None, access_token: str = None) -> List[Issue]:
        
        # Set up authentication if needed        
        headers = {"Authorization": f"Bearer {access_token}"} if access_token else None
        
        # Send the API request    
        api_url = f"{GitHubService.repo_url}{route}"

        print(f"Sending request to {api_url}")
        api_params = {
            "labels": params.label,
            "state": params.state
        }

        response = get(api_url, params=api_params, headers=headers)

        issues_list: List[Issue] = []

        if response.status_code == 200:
            issues = response.json()
            # Process the retrieved issues        
            for issue in issues:
                issues_list.append(Issue(id=issue['number'], title=issue['title'], body=issue['body']))     
        else:
            print(f"Error: {response.status_code} - {response.text}")           
        
        return issues_list    

    # Get issues from a repo
    @staticmethod
    def get_issues(repo: str, label: str, state: str = 'open', access_token: str = None) -> List[Issue]:
        """Get the repo issues."""    
        
        # Define the query parameters
        params = Parameter(label=label, state=state)       
        
        route = f"{repo}/{GitHubService.issues_route}"

        # Send the API request    
        issues_list = GitHubService.get_github_issues(route, params, access_token)
        
        return issues_list


    # get an issue by id
    @staticmethod
    def get_issue_by_id(repo: str, issue_number: str, access_token: str = None) -> Issue:
        """Get the repo issue by issue_number."""

        route = f"{repo}/{GitHubService.issues_route}/{issue_number}"
        
        issue = GitHubService.get_github_issues(route, params=None, access_token=access_token)

        return issue[0] if issue else None


    # post a comment to a GitHub issue
    @staticmethod
    def post_issue_comment(repo: str, issue_number: str, comment: str, access_token: str = None) -> bool:
        
        data = {'body': comment}
        route = f"{repo}/{GitHubService.issues_route}/{issue_number}/{GitHubService.comments_route}"
        print(f"Posting comment to {route}")
        # Set up authentication if needed        
        headers = {"Authorization": f"Bearer {access_token}"} if access_token else None
        api_url = f"{GitHubService.repo_url}{route}"
        response = post(api_url, json=data, headers=headers)        
        
        if response.status_code == 201:
            return True
        else:
            return False